### PR TITLE
AI: add basic enemy

### PR DESCRIPTION
**Pull request by AdameBf**

### CODE REVIEW

File: enemy-test/Makefile [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+$(PROG).gb: $(_OBJ)
+	$(CC) -o $(PROG).gb $^
+
+%.o: %.c
+	$(CC) -o $@ -c $<
+
+assets:
+	$(IMG2GB) tileset \
+		--output-c-file=src/dummy-tileset.c \
+		--output-header-file=src/dummy-tileset.h \
```
Comment (ChatPion) [inline enemy-test/Makefile:27]: FIXME: The output files are wrong, should be `dummy.sprites.c` (according to your code)

File: enemy-test/src/enemy_monster.c [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+	unit->enemy_sprite_r = enemy_sprite_r;
+	unit->sprite_id = s_id;
+	unit->attack_type = attack_type;
+	unit->max_health = hp;
+	unit->health = hp; // Enemies always start full HP.
+	unit->damage = damage;
+	unit->frames_between_attacks = frames_between_attacks;
+	unit->frames_until_next_attack = frames_between_attacks; // There is one full cycle before the enemy starts behaving normally. Might want to be able to configure that.
+}
+
```
Comment (ChatPion) [inline enemy-test/src/enemy_monster.c:18]: We may consider using factory functions to use this one, like

```C
void create_scientist(some params) {
    init_enemy(parameters);
}
```

File: enemy-test/src/enemy_monster.c [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+	unit->xpos = xpos;
+	unit->ypos = ypos;
+	
+	// Initialize left sprite
+	set_sprite_tile(unit->sprite_id, unit->enemy_sprite_l);
+	move_sprite(unit->sprite_id, xpos, ypos);
+	
+	// Initialize right sprite
+	set_sprite_tile(unit->sprite_id + 1, unit->enemy_sprite_r);
+	move_sprite(unit->sprite_id + 1, xpos + 8, ypos);
```
Comment (ChatPion) [inline enemy-test/src/enemy_monster.c:31]: FIXME: Wrong offsets, the gameboy automatically adds (8, 16) to sprites, you need to offset that

File: enemy-test/src/enemy_monster.c [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+	UINT8 former_ypos = unit->ypos; // Retain former y pos for the blinking animation
+	
+	while (blinking_cycles < 10)
+	{
+		wait_vbl_done();
+		wait_vbl_done();
+		wait_vbl_done();
+		wait_vbl_done();
+		wait_vbl_done();
+		wait_vbl_done();
```
Comment (ChatPion) [inline enemy-test/src/enemy_monster.c:49]: FIXME: Don't do that! That will block the _entire_ game!

File: enemy-test/src/enemy_monster.c [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+		
+		if (cycle_state) // Disappears
+		{
```
Comment (ChatPion) [inline enemy-test/src/enemy_monster.c:52]: FIXME: Remain consistent: `if (cycle_state) {` :)
(and do that everywhere else please)

File: enemy-test/src/enemy_monster.c [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+			move_sprite(unit->sprite_id, unit->xpos, 0);
+			move_sprite(unit->sprite_id+1, unit->xpos+8, 0);
+			cycle_state = 0;
+		}
+		else
+		{
+			move_sprite(unit->sprite_id, unit->xpos, former_ypos);
+			move_sprite(unit->sprite_id+1, unit->xpos+8, former_ypos);
```
Comment (ChatPion) [inline enemy-test/src/enemy_monster.c:60]: FIXME: You should definitely wrap `move_sprite` in one custom function.

File: enemy-test/src/enemy_monster.c [hunk 7]
```diff
@@ -0,0 +1,114 @@
...
+			cycle_state = 1;
+			blinking_cycles++;
+		}
+	}
+	
+	// Enemy disappears
+	move_sprite(unit->sprite_id, 168, 0);
+	move_sprite(unit->sprite_id + 1, 168, 0);
```
Comment (ChatPion) [inline enemy-test/src/enemy_monster.c:68]: We'll need to have a "sprite_id" pool, and release this sprite_id at this time

File: enemy-test/src/enemy_monster.c [hunk 8]
```diff
@@ -0,0 +1,114 @@
...
+			unit->health += amount;
+		}
+		else { // Cap it to max health 
+			unit->health = unit->max_health;
+		}
+	}
+}
+
+// Enemy unit launches its attack #TODO: handle MELEE/PROJECTILE
+void enemy_attack(ENEMY* unit) {
```
Comment (ChatPion) [inline enemy-test/src/enemy_monster.c:98]: This signature does not make any sense: you're supposed to attack _someone_! (I do agree that we'll need to define common structs real soon_ish_)

File: enemy-test/src/enemy_test.c [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+		
+		// Proceed by one frame and check whether the enemy can attack
+		basic.frames_until_next_attack--;
+		if (!basic.frames_until_next_attack) {
+			// For testing purposes: regenerate 1 HP before each self-attack, while alive.
+			enemy_hp_regen(&basic, 1);
+			
+			// Attack and reset frame countdown
+			enemy_attack(&basic);
+			basic.frames_until_next_attack = basic.frames_between_attacks;
```
Comment (ChatPion) [inline enemy-test/src/enemy_test.c:54]: FIXME: That definitely should be in enemy.c

**Comment by ChatPion** [comment]
Btw, next name your branch according to your feature. This will remind you _not_ to reuse branches

**Comment by ChatPion** [comment]
Note: before merging, we could rebase and remove the ugly merge commit

File: enemy-test/src/enemy_monster.c (hunk 8 above)
Comment (AdameBf) [reply enemy-test/src/enemy_monster.c:98]: My reasoning here is that enemies will necessarily attack the player. Would you like me to rename the function to `enemy_attack_player`? (even if self-attacking exists in the code)

File: enemy-test/src/enemy_test.c (hunk 9 above)
Comment (AdameBf) [reply enemy-test/src/enemy_test.c]: Fixed

File: enemy-test/src/enemy_monster.c [hunk 10]
```diff
@@ -24,71 +24,58 @@ void display_enemy(ENEMY* unit, UINT8 xpos, UINT8 ypos) {
 	
 	// Initialize left sprite
 	set_sprite_tile(unit->sprite_id, unit->enemy_sprite_l);
-	move_sprite(unit->sprite_id, xpos, ypos);
+	move_sprite(unit->sprite_id, xpos - 8, ypos - 16);
 	
 	// Initialize right sprite
 	set_sprite_tile(unit->sprite_id + 1, unit->enemy_sprite_r);
-	move_sprite(unit->sprite_id + 1, xpos + 8, ypos);
+	move_sprite(unit->sprite_id + 1, xpos, ypos - 16);
```
Comment (ChatPion) [inline enemy-test/src/enemy_monster.c]: I probably should have been clearer, but you actually need to _add_ your offset (the gameboy expects the origins of sprites to be at the bottom-right corner)

File: enemy-test/src/enemy_monster.c (hunk 5 above)
Comment (AdameBf) [reply enemy-test/src/enemy_monster.c]: Consistency changes done

File: enemy-test/src/enemy_monster.c [hunk 11]
```diff
@@ -24,71 +24,58 @@ void display_enemy(ENEMY* unit, UINT8 xpos, UINT8 ypos) {
...
-		else
-		{
+		} else {
 			move_sprite(unit->sprite_id, unit->xpos, former_ypos);
 			move_sprite(unit->sprite_id+1, unit->xpos+8, former_ypos);
 			cycle_state = 1;
 			blinking_cycles++;
 		}
-	}
+	}*/
```
Comment (ChatPion) [inline enemy-test/src/enemy_monster.c]: This definitely isn't what I would call fixing something.
You need to add something akin to a countdown.

File: enemy-test/src/enemy_monster.c (hunk 10 above)
Comment (AdameBf) [reply enemy-test/src/enemy_monster.c]: Yup, I realized shortly after committing. Fixed.

File: enemy-test/src/enemy_monster.c [hunk 12]
```diff
@@ -14,49 +14,58 @@ void init_enemy(ENEMY* unit, UINT8 enemy_sprite_l, UINT8 enemy_sprite_r, UINT8 s
...
-			blinking_cycles++;
-		}
-	}*/
+	// Store current position unless it's out of range
+	if (xpos < 170 && ypos < 170)
+	{
+		unit->xpos = xpos;
+		unit->ypos = ypos;
+	}
+}
```
Comment (ChatPion) [inline enemy-test/src/enemy_monster.c:43]: FIXME: Won't work with big rooms. Not being visible doesn't mean not existing!

File: enemy-test/src/enemy_monster.c [hunk 13]
```diff
@@ -14,49 +14,58 @@ void init_enemy(ENEMY* unit, UINT8 enemy_sprite_l, UINT8 enemy_sprite_r, UINT8 s
...
+		case 13:
+		case 25:
+		case 37:
+			move_enemy(unit, 200, 200);
+			break;
+		case 7: // Make the enemy reappear temporarily
+		case 19:
+		case 31:
+		case 43:
+			move_enemy(unit, unit->xpos, unit->ypos); // This will work because move_enemy does not update (x, y) if it's equal to (200, 200)
```
Comment (ChatPion) [inline enemy-test/src/enemy_monster.c:58]: See comment above

File: enemy-test/src/enemy_monster.c (hunk 12 above)
Comment (ChatPion) [reply enemy-test/src/enemy_monster.c:43]: (You can safely assume that enemies further than (220, 200) will never be visible however)

File: enemy-test/src/enemy_monster.c [hunk 14]
```diff
@@ -14,49 +14,58 @@ void init_enemy(ENEMY* unit, UINT8 enemy_sprite_l, UINT8 enemy_sprite_r, UINT8 s
...
-// Play death sequence (blinking), then make the enemy disappear
-void enemy_death(ENEMY* unit) {
-	// Blinking animation - disabled
-	/*
-	UINT8 blinking_cycles = 0;
-	UINT8 cycle_state = 1;
+void move_enemy(ENEMY* unit, UINT8 xpos, UINT8 ypos) {
+	// Moves enemy unit to (x,y)
+	move_sprite(unit->sprite_id, xpos + 8, ypos + 16);
+	move_sprite(unit->sprite_id + 1, xpos + 16, ypos + 16);
```
Comment (ChatPion) [inline enemy-test/src/enemy_monster.c:35]: We'll have to use a `#define` for these constants

**Pull request merged**